pkg/operator/quorumguardcontroller/quorumguardcontroller.go

```diff
@@ -0,0 +1,248 @@
+	err := c.ensureEtcdGuard(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "QuorumGuardControllerDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "Error",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("QuorumGuardControllerUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient,
+		v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "QuorumGuardControllerDegraded",
+			Status: operatorv1.ConditionFalse,
+			Reason: "AsExpected",
+		}))
+	return updateErr
+}
+
+func (c *QuorumGuardController) ensureEtcdGuard(ctx context.Context, recorder events.Recorder) error {
+	haTopologyMode, err := c.isInHATopologyMode()
+	if err != nil {
+		klog.Infof("Failed to validate ha mode %v ", err)
+		return err
+	}
+	if !haTopologyMode {
+		return nil
+	}
+
+	replicaCount, err := c.getMastersReplicaCount(ctx)
+	if err != nil {
+		return err
+	}
+
+	if err := c.ensureEtcdGuardDeployment(ctx, replicaCount, recorder); err != nil {
+		return err
+	}
+
+	if err := c.ensureEtcdGuardPdbDeployment(ctx, recorder); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ensureEtcdGuardDeployment if etcd quorum guard deployment doesn't exist - add it, else skip
+func (c *QuorumGuardController) ensureEtcdGuardDeployment(ctx context.Context, replicaCount int32, recorder events.Recorder) error {
+	_, err := c.kubeClient.AppsV1().Deployments(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdQuorumGuardDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyDeployment(replicaCount, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// ensureEtcdGuardDeployment if etcd quorum guard pdb deployment doesn't exist - add it, else skip
+func (c *QuorumGuardController) ensureEtcdGuardPdbDeployment(ctx context.Context, recorder events.Recorder) error {
+	_, err := c.kubeClient.PolicyV1beta1().PodDisruptionBudgets(operatorclient.TargetNamespace).Get(ctx,
+		EtcdGuardDeploymentName, metav1.GetOptions{})
+	switch {
+	case apierrors.IsNotFound(err):
+		recorder.Eventf("NoEtcdQuorumGuardPDBDeployment", "%s was not found, creating one", EtcdGuardDeploymentName)
+		return c.applyPdbDeployment(ctx, recorder)
+	case err != nil:
+		klog.Warningf("failed to get %s deployment err %v", EtcdGuardDeploymentName, err.Error())
+		return err
+	default:
+		return nil
+	}
+}
+
+// applyDeployment applies new etcd quorum guard deployment
+func (c *QuorumGuardController) applyDeployment(replicaCount int32, recorder events.Recorder) error {
```

Comment:
optional nit: if we only need a client can we drop this as a method and just pass the client?